lib/signs/bus.ex

```diff
@@ -97,18 +133,70 @@ defmodule Signs.Bus do
-  # defp display_route(%{route_id: "743"}), do: "SL3"
-  # defp display_route(%{route_id: "742"}), do: "SL2"
-  # defp display_route(%{route_id: "741"}), do: "SL1"
-  # defp display_route(%{route_id: "77", headsign: "North Cambridge"}), do: "77A"
-  # defp display_route(%{route_id: "2427", stop_id: "185", headsign: headsign}) do
-  #  cond do
-  #    String.starts_with?(headsign, "Ashmont") -> "27"
-  #    String.starts_with?(headsign, "Wakefield Av") -> "24"
-  #    true -> "2427"
-  #  end
-  # end
-  # defp display_route(%{route_id: route_id}), do: route_id
+  defp display_route(%{route_id: "746"}), do: "SLW"
+  defp display_route(%{route_id: "749"}), do: "SL5"
+  defp display_route(%{route_id: "751"}), do: "SL4"
+  defp display_route(%{route_id: "743"}), do: "SL3"
+  defp display_route(%{route_id: "742"}), do: "SL2"
+  defp display_route(%{route_id: "741"}), do: "SL1"
+  defp display_route(%{route_id: "77", headsign: "North Cambridge"}), do: "77A"
+
+  defp display_route(%{route_id: "2427", stop_id: "185", headsign: headsign}) do
+    cond do
+      String.starts_with?(headsign, "Ashmont") -> "27"
+      String.starts_with?(headsign, "Wakefield Av") -> "24"
+      true -> "2427"
+    end
+  end
+
+  defp display_route(%{route_id: route_id}), do: route_id
+
+  defp prediction_minutes(prediction, current_time) do
+    Timex.diff(prediction.departure_time, current_time, :minutes)
+  end
+
+  defp prediction_key(prediction) do
+    Map.take(prediction, [:stop_id, :route_id, :vehicle_id, :direction_id])
+  end
+
+  defp format_predictions([first, second], current_time) do
+    same = second && first.route_id == second.route_id && first.headsign == second.headsign
+
+    for prediction <- [first, second] do
+      if prediction do
+        route = display_route(prediction)
+        # If both predictions are for the same route, pad both lines based on the second line's
+        # time string. That may use up extra space on the top line, but it means we'll choose
+        # the same abbreviation for both, so the text will line up nicely.
+        time =
+          String.pad_leading(
+            format_time(prediction, current_time),
+            String.length(format_time(if(same, do: second, else: prediction), current_time))
+          )
+
+        dest_max = @line_max - String.length(route) - String.length(time) - 2
+
+        # Choose the longest abbreviation that will fit within the remaining space.
+        dest =
+          (PaEss.Utilities.headsign_abbreviations(prediction.headsign) ++ [prediction.headsign])
+          |> Enum.filter(&(String.length(&1) <= dest_max))
+          |> Enum.max_by(&String.length/1, fn ->
+            Logger.warn("No abbreviation for headsign: #{inspect(prediction.headsign)}")
+            String.slice(prediction.headsign, 0, dest_max)
+          end)
+          |> String.pad_trailing(dest_max)
+
+        "#{route} #{dest} #{time}"
+      else
+        ""
+      end
+    end
+  end
```

Comment:
Ohhh gotcha, I see now. Yeah it was the part of the code for calculating the time string mainly that I found a bit hard to read so I didn't realize the intention behind doing it this way.

It also seems to me like the for comprehension may not be necessary given that we have three pretty well defined cases here:
- [prediction, nil]
- [prediction, prediction] when destination is the same
- [prediction, prediction]

We could maybe explore a solution that expresses that a bit more explicitly? Personally I think I would find that a bit easier to follow, but let me know what you think about that.

Reply:
I suspect we'll get there, especially since the mezzanine case will probably require some very similar rendering code, but without being part of a linked pair. And there's probably a general case to find there, but I'm hesitant to make too many assumptions before seeing that code take shape.

I updated the comment to call out the quirky behavior more explicitly. Is it ok to move forward with this for now, with the understanding that it will almost certainly morph further as we build out the cases?

Reply:
Fair point. Sounds good to me! And thank you for updating the comment, definitely helps clarify what is going on there!